### PR TITLE
EP Merge: Added abort for ineligible claim requests

### LINKED ISSUE #3241: EP Merge: Abort for ineligible requests
# Summary
Currently, EP Merge does not abort jobs that should have been ineligible from the start. For example, EP Merge will fail in error if the benefit claim type code of the EP400 is anything but `400SUPP`. This was noticed in production [logs](https://lighthousedi.ddog-gov.com/logs?query=service%3Avro-ee-ep-merge-app%20%22benefit%20claim%20type%20code%20of%22%20&agg_m=count&agg_m_source=base&agg_t=count&cols=host%2Cservice&context_event=AZDlfPTPAAAkbf8CiFmSnQAN&fromUser=true&fullscreen_end_ts=1722023458716&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1721418658716&fullscreen_widget=3030054547686168&messageDisplay=inline&refresh_mode=sliding&screenId=unv-vap-yeb&screenName=ep-merge-employee-experience-vro-platform&storage=hot&stream_sort=time%2Cdesc&view=spans&viz=&from_ts=1719431521607&to_ts=1722023521607&live=true) when jobs were failing at the `GET_EP400_CLAIM` state. 

## Acceptance Criteria
EP Merge should be updated to include a new ABORTED end state, where any claims that should not be merged due to ineligible criteria: 
* Pending claim details response is OK but contains no claim details <sup>1</sup>
* Pending claim details end product code is not populated <sup>1</sup>
* Pending claim is not open when the request is received <sup>2</sup>
* EP400 claim details response is OK but contains no claim details <sup>1</sup>
* 🆕 EP400 claim is not open when the request is received before getting claim contentions <sup>2</sup>
* EP400 claim details end product code is not populated <sup>1</sup>
* EP400 claim details benefit claim type code is not populated <sup>1</sup>
* EP400 claim details benefit claim type code is not `400SUPP` 
* EP400 claim does not have any contentions after 60s <sup>3</sup>
* Pending Ep is no longer open after gathering EP400 contentions <sup>2</sup>
* 🆕 EP 400 is no longer open after gathering EP400 contentions <sup>2</sup> 

<sup>1</sup> - This is unlikely to happen in any deployed environment, but rather a result of not having a good enough mock service to represent actual data.
<sup>2</sup> - This is a mechanism to address the race of EP Merge to begin processing vs the claim being closed elsewhere, bottom line = closed claims are not eligible.
<sup>3</sup> - This will only happen if there is a 60 second slow down in the time it takes the claim to have populated contentions, if this is the case the slow down will not be detectable by EP Merge or VRO downstream services, bottom line = No contentions on EP400 means claim is not eligible to merge

### PR DESCRIPTION
**Pull request by dfitchett**
## What was the problem?

Currently, EP Merge does not abort jobs that should have been ineligible from the start. 

Associated tickets or Slack threads:
- #3241 

## How does this fix it?

Adds an ABORTING and ABORTED state. See acceptance criteria in the ticket for when jobs should be aborted.

This is the old state machine:

EP_process (1)

And this is the new implementation:
EP_process

## How to test this PR
- pull code
- Unit Test:
  - from /domain-ee/ee-ep-merge-app run `pytest`
- Integration Tests:
  - run [base platform](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Docker-Compose#platform-base)
  - from /domain-ee/ee-ep-merge-app run `pytest ./integration`
- End to End tests:
  - run this [CI](https://github.com/department-of-veterans-affairs/abd-vro/actions/workflows/ee-ep-merge-end-to-end.yml) action to verify successful integration test
  - run end to end tests locally:
    - `export BIP_CLAIM_URL=mock-bip-claims-api:20300`
    - run bgs and bip services and mocks:
      - `COMPOSE_PROFILES=“all” ./gradlew dockerComposeUp`
      - `COMPOSE_PROFILES="bip,bgs" ./gradlew -p mocks :dockerComposeUp`
    - `./gradlew :domain-ee:ee-ep-merge-app:endToEndTest`

### CODE REVIEW

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 1]
```diff
@@ -490,6 +505,9 @@ def make_request(
     def has_error(self) -> bool:
         return self.job.state == JobState.COMPLETED_ERROR  # type: ignore[no-any-return]
 
+    def is_aborted(self) -> bool:
+        return self.job.state == JobState.ABORTED  # type: ignore[no-any-return]
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]: Oh, I see why the ignore comment is here, to placate the pedantry of `__eq__` potentially returning non-bool. Maybe it's better to just `bool()`ify the value? (I got the idea from this [SO post](https://stackoverflow.com/questions/75864807/mypy-complaining-about-no-any-return-rule-being-violated-for-obvious-boolean-e).)

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 2]
```diff
@@ -280,17 +276,30 @@ def on_get_ep400_contentions(self, event, pending_contentions_response=None):
...
-        request = get_claim.Request(claim_id=self.job.pending_claim_id)
+        self.__get_open_claim(self.job.pending_claim_id, 'Pending')
+        self.send(event=event, pending_contentions_response=pending_contentions_response, ep400_contentions_response=ep400_contentions_response)
+
+    @running_check_ep400_is_open.enter
+    def on_check_ep400_claim_is_open(self, event, pending_contentions_response, ep400_contentions_response):
+        self.__get_open_claim(self.job.ep400_claim_id, 'EP400')
+        self.send(event=event, pending_contentions_response=pending_contentions_response, ep400_contentions_response=ep400_contentions_response)
+
+    def __get_open_claim(self, claim_id: int, claim_type: str) -> ClaimDetail | None:
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:287]: I'm not the biggest fan of `__private` methods unless it's guarding against a real risk of outside code dangerously mucking with internal state. What do you think?

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py [hunk 3]
```diff
@@ -508,8 +526,14 @@ def has_contentions_for_merge(
 
     def add_job_error(self, message: str) -> None:
         errors = {'state': self.job.state, 'error': message}
-        logging.warning(f'event=jobError job_id={self.job.job_id} error={jsonable_encoder(errors)}')
-        self.job.error(errors)
+
+        # if job was resumed from either of these points, the contentions were already moved over, and the job should end in error instead of aborting
+        if self.main_event == Workflow.RESUME_CANCEL_EP400 or self.main_event == Workflow.RESUME_ADD_NOTE:
```
Comment (nanotone) [inline domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py]:
```suggestion
        if self.main_event in (Workflow.RESUME_CANCEL_EP400, Workflow.RESUME_ADD_NOTE):
```

**Review by nanotone — APPROVED** [review]
This mostly looks good to me -- "mostly" not because I have any strong objections in the couple of suggestions I left (they're all very minor) but because I focused my attention on `ep_merge_machine.py` and glanced over the other files more quickly since there's a bunch of boilerplate changes. Let me know if you want scrutiny on some other part of this PR.

File: domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py (hunk 2 above)
Comment (dfitchett) [reply domain-ee/ee-ep-merge-app/src/python_src/service/ep_merge_machine.py:287]: Yeah i suppose methods are never truly private anyhow, and in this case the method is not changing state, so I'm cool with changing it. Encapsulation isn't strictly enforced in python and I haven't been marking all the others as private up to this point either 😅

**Pull request merged**